src/tfs/tfs.c

```diff
@@ -576,335 +460,245 @@ static void remove_extent_from_file(fsfile f, extent ex, merge m)
-    merge m_meta = allocate_merge(fs->h, closure(fs->h, filesystem_write_meta_complete, q, ish));
-    merge m_data = allocate_merge(fs->h, closure(fs->h, filesystem_write_data_complete,
-                                                 f, t, q, m_meta, apply_merge(m_meta)));
-
-    /* hold data merge open until all extent operations have been initiated */
-    status_handler sh = apply_merge(m_data);
-    do {
-        /* detect and fill any hole before extent (or to end) */
-        u64 limit = node != INVALID_ADDRESS ? node->r.start : q.end;
-        if (curr < limit) {
-            range hole = irange(curr, limit);
-            range fill = range_intersection(q, hole);
-
-            /* XXX optimization: check for a preceding extent and
-               inflate if possible */
-
-            /* just doing min-sized extents for now */
-            s64 remain = range_span(fill);
-
-            do {
-                /* create_extent will allocate a minimum of pagesize */
-                u64 length = MIN(MAX_EXTENT_SIZE, remain);
-                range r = irange(curr, curr + length);
-                extent ex = fs_new_extent(f, r, false, m_meta);
-                if (ex == INVALID_ADDRESS) {
-                    msg_err("failed to create extent\n");
-                    goto fail;
-                }
-                tfs_debug("   writing new extent %R\n", r);
-                fs_write_extent(f->fs, b, m_data, q, &ex->node);
-                curr += length;
-                remain -= length;
-            } while (remain > 0);
-        }
-
-        if (node != INVALID_ADDRESS) {
-            /* overwrite any overlap with extent */
-            range i = range_intersection(q, node->r);
-            if (range_span(i)) {
-                tfs_debug("   updating extent at %R (intersection %R)\n", node->r, i);
-                fs_write_extent(f->fs, b, m_data, q, node);
-                extent e = (extent)node;
-                if (e->uninited) {
-                    if (!extents) {
-                        extents = table_find(t, sym(extents));
-                        if (!extents) {
-                            msg_err("no extents in tuple %t\n", t);
-                            goto fail;
-                        }
-                    }
-                    symbol offs = intern_u64(node->r.start);
-                    tuple extent_tuple = table_find(extents, offs);
-                    if (!extent_tuple) {
-                        msg_err("failed: can't find extent tuple\n");
-                        goto fail;
-                    }
-                    if (table_find(extent_tuple, sym(uninited))) {
-                        tfs_debug("   removing uninited flag\n");
-                        table_set(extent_tuple, sym(uninited), 0);
-                        filesystem_write_eav(f->fs, extent_tuple, sym(uninited), 0,
-                                apply_merge(m_meta));
-                    }
-                    e->uninited = false;
```

Comment:
This block of code, whose purpose is to remove the "uninited" flag from an extent being filled with user data, seems to be missing in the new code. This causes file data to always read as zeros for files that have been `fallocate()`d, even after the files have been written to.